presto-main/src/main/java/com/facebook/presto/resourcemanager/DistributedTaskInfoResource.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.resourcemanager;
+
+import com.facebook.presto.execution.TaskId;
+import com.facebook.presto.server.BasicQueryInfo;
+import com.facebook.presto.spi.QueryId;
+
+import javax.annotation.security.RolesAllowed;
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.container.Suspended;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriInfo;
+
+import java.util.Optional;
+
+import static com.facebook.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static com.facebook.presto.server.security.RoleType.ADMIN;
+import static com.facebook.presto.server.security.RoleType.USER;
+import static java.util.Objects.requireNonNull;
+import static javax.ws.rs.core.Response.Status.NOT_FOUND;
+
+@Path("/v1/taskInfo")
+@RolesAllowed({USER, ADMIN})
+public class DistributedTaskInfoResource
+{
+    private final ResourceManagerClusterStateProvider clusterStateProvider;
+    private final ResourceManagerProxy proxyHelper;
+
+    @Inject
+    public DistributedTaskInfoResource(
+            ResourceManagerClusterStateProvider clusterStateProvider,
+            ResourceManagerProxy proxyHelper)
+    {
+        this.clusterStateProvider = requireNonNull(clusterStateProvider, "clusterStateProvider is null");
+        this.proxyHelper = requireNonNull(proxyHelper, "proxyHelper is null");
+    }
+
+    @GET
+    @Path("{taskId}")
+    public void getTaskInfo(@PathParam("taskId") TaskId taskId,
+            @Context UriInfo uriInfo,
+            @Context HttpServletRequest servletRequest,
+            @Suspended AsyncResponse asyncResponse)
+            throws WebApplicationException
+    {
+        proxyTaskInfoResponse(servletRequest, asyncResponse, uriInfo, taskId);
+    }
+
+    private void proxyTaskInfoResponse(HttpServletRequest servletRequest, AsyncResponse asyncResponse, UriInfo uriInfo, TaskId taskId)
+    {
+        QueryId queryId = taskId.getQueryId();
+        Optional<BasicQueryInfo> queryInfo = clusterStateProvider.getClusterQueries().stream()
+                .filter(query -> query.getQueryId().equals(queryId))
+                .findFirst();
```

Comment:
would there be duplicates here?

Reply:
There shouldn't be:
* `getClusterQueries()` returns the list of all active and completed queries.
* These `activeQueries` and `completedQueries` are stored in `Map<QueryId, Query>`s, which will guarantee the uniqueness of the keys,
* We can also assume the uniqueness of the values too, because of their 1:1 relationship with the keys.
* Lastly, it shouldn't be possible for a query to be in both maps, and hence a duplicate, because of the remove logic in the last code block

https://github.com/prestodb/presto/blob/89e259a8b3418e9099f19d909b85e600667295bc/presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java#L255-L262

https://github.com/prestodb/presto/blob/89e259a8b3418e9099f19d909b85e600667295bc/presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java#L458-L462

https://github.com/prestodb/presto/blob/89e259a8b3418e9099f19d909b85e600667295bc/presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java#L384-L387

https://github.com/prestodb/presto/blob/89e259a8b3418e9099f19d909b85e600667295bc/presto-main/src/main/java/com/facebook/presto/resourcemanager/ResourceManagerClusterStateProvider.java#L419-L422

Reply:
curious then why need findFirst() , find() is clearer in this case

Reply:
Switched to `.findAny()`